### DISCUSSION TITLE
Engines fixed directories

### DISCUSSION THREAD

**Discussion by Octal450** [JSBSim-Team/jsbsim#220] [Support requests]
Hi - I am wondering if it can be known - why does the engine MUST be at Aircraft_DIR/Engines/myengine.xml?

Why Cannot be in a place I define - like everything else?

I would ask if it can be modified so that I can define my own path to engine.

Kind Regards,
Josh
PS: I am very tired at the moment, please let me know if I need more clarify.

**Comment by ghost** [comment]
This has been bothering me as well -- why not use common system of paths with `<system include="blabla">` and such? Then engines and thrusters would be in the JSBSim subdirectory of the aircraft, together with all other systems.

For example, in my plane the whole logic model is under https://gitlab.com/mdanil/Tu-144/tree/master/Tu-144-jsbsim -- but engines they have to be https://gitlab.com/mdanil/Tu-144/tree/master/Engines -- that doesn't make any sense...

**Comment by Octal450** [comment]
Well it's actually an even simpler fix. Just add support for custom path define, just like every other file, system, aerodynamic, flight_control, etc

Kind Regards,
Josh

**Comment by bcoconni** [comment]
JSBSim provides the ability to modify the folder in which the engines are loaded with the C++ function `JSBSim::FGFDMExec::SetEnginePath`  which is documented in [Developers docs](https://jsbsim-team.github.io/jsbsim/classJSBSim_1_1FGFDMExec.html#a407040feba3cfb275c6cc19e65f4cee5).

**Comment by Octal450** [comment]
Hi Bertrand,
Perhaps I am just not seeing it, but how do I use this from the XML side?
Thanks.

Kind Regards,
Josh

**Comment by bcoconni** [comment]
Sorry, I was in a rush the other day and my answer lacked some details.

Actually there is no such option as `include=...` neither for `<system>` nor for any other element. The option you are referring to is `file=...` and it is currently used to specify the file name of (or the path to) the engines definition. If the attribute `file=...` is provided with a relative path then JSBSim will look for files located under `Engines` or more generally under the engines default path (which is set to `engine` by default). This is the way that is almost always used in FlightGear to specify the path to engines definition files.

However, if the attribute `file=...` is provided with an absolute path then JSBSim can theoretically read a file that is located anywhere in the file system. This feature is available in FlightGear for a long time. If it does not currently work then it is a bug.

The last option is to modify the engine default path but this must be done before the aircraft model is loaded. As I mentioned earlier, there is already an option for that which is the C++ method `FGFDMExec::SetEnginePath`. For that option to be made available, you'd better ask the FlightGear devs to expose an option in FlightGear that allows to alter the JSBSim default engine path.

**Comment by ghost** [comment]
Yes of course I meant `<system file="">`. I'm already using `<engine file="">` and it does not pick up engines from anything but the `Engines/` directory.

**Comment by bcoconni** [comment]
@mike402 
Then your options are:
1. Provide an absolute path to `<engine file=...>`
2. Ask the FlightGear devs to provide an option to alter the default path `Engines` via the C++ method `FGFDMExec::SetEnginePath`.
